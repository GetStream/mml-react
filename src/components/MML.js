import parseXml from "@rgrove/parse-xml";
import ReactPlayer from "react-player";
import React from "react";
import { sanitizeUrl } from "@braintree/sanitize-url";
import { MMLMarkdown } from "./MMLMarkdown";
import DatePicker from "react-datepicker";
import { Carousel } from "./Carousel";
import { Loader } from "./Loader";
import { Error } from "./Error";
import { Success } from "./Success";

export class MML extends React.PureComponent {
  static defaultProps = {
    loaderComponent: Loader,
    errorComponent: Error,
    successComponent: Success
  };

  constructor(props) {
    super(props);

    this.state = { error: "", loading: false, success: "" };
  }

  handleSubmit = async event => {
    event.preventDefault();
    const data = [];
    const pairs = Object.keys(this.state).map((key, index) => {
      return { name: key, value: this.state[key] };
    });
    data.push(...pairs);

    console.log("loading...");
    this.setState({ loading: true, error: "", success: "" });
    try {
      await this.props.onAction(data);
      console.log("done");
      this.setState({ loading: false, error: "", success: "submitted" });
    } catch (e) {
      console.log("error");
      this.setState({ loading: false, error: "something is broken" });
      console.log("e", e);
    }
  };

  handleAction = (attr, event) => {
    console.log("handling action", attr, event);

    if (attr.url && attr.url.length) {
      window.location.href = sanitizeUrl(attr.url);
    } else {
      const data = {};
      data[attr.name] = attr.value || event.target.value;
      console.log("setting state", data);
      this.setState(data);
    }
  };

  handleDateChange = (attr, selectedDate) => {
    const data = {};
    data[attr.name] = selectedDate;
    this.setState(data);
  };

  render() {
    const that = this;

    /*
     * anything that changes state has a name attribute (select, text input etc.)
     */
    let mmlName = "";

    function filterDate(date) {
      return date;
    }

    function mmlToHTML(nodes) {
      const html = [];

      for (let n of nodes) {
        let children = "";
        if (n.children) {
          children = mmlToHTML(n.children);
        }

        if (n.name === "column") {
          const width = n.attributes.width ? n.attributes.width : 12;
          const offset = n.attributes.offset ? n.attributes.offset : 0;
          let classNames = `mml-col-${width}`;
          if (offset) {
            classNames = classNames + ` mml-offset-${offset}`;
          }

          const align = n.attributes.align ? n.attributes.align : "left";
          classNames = classNames + ` mml-align-${align}`;

          html.push(<div className={classNames}>{children}</div>);
        } else if (n.name === "icon") {
          html.push(<i class="material-icons">{n.attributes.name}</i>);
        } else if (n.name === "card") {
          html.push(<div class="mml-card">{children}</div>);
        } else if (n.name === "buttonlist") {
          html.push(<div class="mml-selectlist">{children}</div>);
        } else if (n.name === "row") {
          html.push(<div className="mml-row">{children}</div>);
        } else if (n.name === "overflow") {
          html.push(
            <div class="dropdown">
              <button
                class="btn btn-secondary dropdown-toggle"
                type="button"
                id="dropdownMenu2"
                data-toggle="dropdown"
                aria-haspopup="true"
                aria-expanded="false"
              >
                ...
              </button>
              <div class="dropdown-menu" aria-labelledby="dropdownMenu2">
                {children}
              </div>
            </div>
          );
        } else if (n.name === "mml") {
          console.log("n.attributes.name", n.attributes.name);
          mmlName = n.attributes.name;
          html.push(
            <div className={`mml-container mml-${that.props.layout}`}>
              <form onSubmit={that.handleSubmit}>
                {children}
                <Loader loading={that.state.loading} />
                <Success success={that.state.success} />
                <Error error={that.state.error} />
              </form>
            </div>
          );
        } else if (n.name === "md") {
          html.push(<MMLMarkdown source={n.children[0].text} />);
        } else if (n.name === "text") {
          html.push(<p>{n.children[0].text}</p>);
        } else if (n.type === "text") {
          if (n.text.trim().length > 0) {
            html.push(<p>{n.text}</p>);
          }
        } else if (n.name === "input") {
          html.push(
            <input
              {...n.attributes}
              onChange={that.handleAction.bind(that, n.attributes)}
            />
          );
        } else if (n.name === "separator") {
          html.push(<hr />);
        } else if (n.name === "image") {
          html.push(<img {...n.attributes} />);
        } else if (n.name === "select") {
          html.push(
            <select
              {...n.attributes}
              onChange={that.handleAction.bind(that, n.attributes)}
            >
              {children}
            </select>
          );
        } else if (n.name === "option") {
          html.push(
            <option value={n.attributes.value}>{n.children[0].text}</option>
          );
        } else if (n.name === "carousel") {
          html.push(<Carousel items={children} />);
        } else if (n.name === "item") {
          html.push(<div className="mml-carousel-item">{children}</div>);
        } else if (n.name === "datepicker") {
          html.push(
            <DatePicker
              selected={that.state[n.attributes.name]}
              onChange={that.handleDateChange.bind(that, n.attributes)}
              filterDate={filterDate}
            />
          );
        } else if (n.name === "timepicker") {
          const initialDate = new Date(n.attributes.initial_date);
          html.push(
            <DatePicker
              selected={that.state[n.attributes.name]}
              showTimeSelect
              showTimeSelectOnly
              timeIntervals={15}
              dateFormat="h:mm aa"
              timeCaption="Time"
              onChange={that.handleDateChange.bind(that, n.attributes)}
            />
          );
        } else if (n.name === "video") {
          html.push(
            <ReactPlayer
              className="react-player"
              url={n.attributes.url}
              width="100%"
              controls
            />
          );
        } else if (n.name === "button") {
          // TODO: create a button component that knows how to navigate to a url
          // TODO: fix this
          html.push(
            <button
              className={`mml-btn`}
              type="submit"
              onClick={() => that.handleAction(n.attributes)}
            >
              {n.attributes.value}
            </button>
          );
        } else if (n.type === "document") {
          return children;
        } else {
          console.log("unrecognized element", n);
        }
      }

      return html;
    }
    console.log("mmlName", mmlName);

    const html = mmlToHTML([parseXml(this.props.source)]);
    if (mmlName) {
      this.setState({ mml_name: mmlName });
    }
    console.log("html", html);
    return html;
  }
}
