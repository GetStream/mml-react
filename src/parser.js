import parseXml from '@rgrove/parse-xml'
import { getMMLTags } from './tags'
import { Tree } from './tree'

/**
 * Takes an MML string and returns an MML Tree
 *
 * @param {string} source MML tag string
 *
 *  @returns {Tree} An MML Tree
 */
export function Parse(source) {
  const XMLNodes = SourceToXML(source)
  const tree = XMLtoMMLTree(XMLNodes)
  return tree
}

/**
 * SourceToXML - Takes an MML string and converts it to XML nodes
 *
 * @param {string} source MML tag string
 *
 * @returns {array} an Array of XML nodes
 */
export function SourceToXML(source) {
  // the wrapping MML tags are optional, for parsing simplicity we automatically add them if they are not already there
  if (!~source.indexOf('<mml')) {
    source = `<mml>${source}</mml>`
  }

  // emulate HTML handling of & escaping
  const unescapedAmps = /&(?!amp;|lt;|gt;)/g
  source = source.replace(unescapedAmps, '&amp;')

  // convert the string to XML nodes
  // this library is relatively lightweight and doesn't do a ton of validation
  let XMLNodes
  try {
    XMLNodes = [parseXml(source)]
  } catch (err) {
    let safeErr = err
      .toString()
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#039;')
    XMLNodes = [parseXml('<mml><text>' + safeErr + '</text></mml>')]
  }
  return XMLNodes
}

/**
 * XMLtoMMLTree - Takes an array of XML nodes and converts it into an MML Tree
 *
 * @param {type} XMLNodes an array of XML nodes
 *
 * @returns {MMLTree} The MML tree
 */
export function XMLtoMMLTree(XMLNodes) {
  const tags = getMMLTags()

  let tree

  function convertNodes(nodes) {
    const MMLNodes = []
    for (const n of nodes) {
      let children
      if (n.children) {
        children = convertNodes(n.children)
      }

      // structured way of looking up mml tags...
      let tagName = n.name
      if (n.name === 'mml') {
        tree = new Tree(n, children)
        continue
      }
      // skip the document level element...
      if (n.type === 'document') {
        return children
      }
      if (n.type === 'text') {
        if (n.text.trim().length > 0) {
          tagName = 'text'
        } else {
          // skip empty text nodes
          continue
        }
      }

      const TagClass = tags[tagName]

      if (TagClass) {
        const tag = new TagClass(tagName, n, children)
        MMLNodes.push(tag)
      } else {
        console.log('unrecognized element', tagName, Object.keys(tags))
      }
    }
    return MMLNodes
  }
  convertNodes(XMLNodes)

  return tree
}
