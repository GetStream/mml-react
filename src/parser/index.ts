export * from './parser';
export * from './MMLTag';
export * from './tree';
export * from './converters';
