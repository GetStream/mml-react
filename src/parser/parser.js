import parseXml from '@rgrove/parse-xml';
import { tags } from './tags';
import { Tree } from './tree';

/**
 * SourceToXML - Takes an MML string and converts it to XML nodes
 *
 * @param {string} source MML tag string
 *
 * @returns {array} an Array of XML nodes
 */
export function SourceToXML(source) {
  let src = source;
  // the wrapping MML tags are optional, for parsing simplicity we automatically add them if they are not already there
  if (!src.startsWith('<mml')) src = `<mml>${source}</mml>`;

  // emulate HTML handling of & escaping
  const unescapedAmps = /&(?!amp;|lt;|gt;)/g;
  src = src.replace(unescapedAmps, '&amp;');

  // convert the string to XML nodes
  // this library is relatively lightweight and doesn't do a ton of validation
  return [parseXml(src)];
}

/**
 * XMLtoMMLTree - Takes an array of XML nodes and converts it into an MML Tree
 *
 * @param {type} XMLNodes an array of XML nodes
 *
 * @returns {MMLTree} The MML tree
 */
export function XMLtoMMLTree(XMLNodes) {
  let tree;

  function convertNodes(nodes) {
    const MMLNodes = [];
    for (const n of nodes) {
      let children;
      if (n.children) {
        children = convertNodes(n.children);
      }

      // structured way of looking up mml tags...
      let tagName = n.name;
      if (n.name === 'mml') {
        tree = new Tree(n, children);
        continue;
      }
      // skip the document level element...
      if (n.type === 'document') {
        return children;
      }
      if (n.type === 'text') {
        if (n.text.trim().length > 0) {
          tagName = 'text';
        } else {
          // skip empty text nodes
          continue;
        }
      }

      const TagClass = tags[tagName];

      if (TagClass) {
        const tag = new TagClass(tagName, n, children);
        MMLNodes.push(tag);
      } else {
        console.log('unrecognized element', tagName, Object.keys(tags));
      }
    }
    return MMLNodes;
  }
  convertNodes(XMLNodes);

  return tree;
}

/**
 * Takes an MML string and returns an MML Tree
 *
 * @param {string} source MML tag string
 *
 *  @returns {Tree} An MML Tree
 */
export function Parse(source) {
  const XMLNodes = SourceToXML(source);
  return XMLtoMMLTree(XMLNodes);
}
