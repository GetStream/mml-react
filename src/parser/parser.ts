import parseXml from '@rgrove/parse-xml';

import { ConvertorType } from './converters';
import { MMLTag } from './MMLTag';
import { Tree } from './tree';

/**
 * SourceToXML - Takes an MML string and converts it to XML nodes
 *
 * @param {string} source MML tag string
 *
 * @returns {array} an Array of XML nodes
 */
export function SourceToXML(source: string) {
  let src = source.trim();
  // the wrapping MML tags are optional, for parsing simplicity we automatically add them if they are not already there
  if (!src.startsWith('<mml')) src = `<mml>${source}</mml>`;

  // emulate HTML handling of & escaping
  const unescapedAmps = /&(?!amp;|lt;|gt;)/g;
  src = src.replace(unescapedAmps, '&amp;');

  // convert the string to XML nodes
  // this library is relatively lightweight and doesn't do a ton of validation
  return parseXml(src);
}

function convertNodes(nodes: parseXml.XmlNode[]) {
  return nodes.reduce((acc: MMLTag[], node: parseXml.XmlNode) => {
    const element = node as parseXml.XmlElement;
    let children;
    if (element.children) children = convertNodes(element.children);

    let { name } = element;
    if ((node as parseXml.XmlText).type === 'text') {
      if ((node as parseXml.XmlText).text.trim().length) name = 'text';
      else return acc; // skip empty text elements
    }

    acc.push(new MMLTag(name, node as parseXml.XmlElement | parseXml.XmlText, children));
    return acc;
  }, []);
}

/**
 * XMLtoMMLTree - Takes an array of XML nodes and converts it into an MML Tree
 *
 * @param {type} XMLNodes an array of XML nodes
 *
 * @returns {MMLTree} The MML tree
 */
export function XMLtoMMLTree(document: parseXml.XmlDocument, customConvertors?: Record<string, ConvertorType>) {
  if (!document || !document.children || !document.children.length) throw new Error('bad input');

  const mmlNode = document.children[0] as parseXml.XmlElement;
  if (mmlNode.name !== 'mml') throw new Error('missing mml tag');

  return new Tree(mmlNode, convertNodes(mmlNode.children), customConvertors);
}

/**
 * Takes an MML string and returns an MML Tree
 *
 * @param {string} source MML tag string
 *
 *  @returns {Tree} An MML Tree
 */
export function Parse(source: string, customConvertors?: Record<string, ConvertorType>) {
  const XMLNodes = SourceToXML(source);
  return XMLtoMMLTree(XMLNodes, customConvertors);
}
