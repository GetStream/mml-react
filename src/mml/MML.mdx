---
name: MML
route: /mml
---

import { Playground, Props } from 'docz';
import { MML } from './MML.tsx';

# MML

This component accepts an MML string and renders it as React components.

If there are any stateful tag, it will gather all the inital states and and expose them in [`MMLContext`](/mml-context).

It also accepts custom convertor which allows you to exapand the default tags and add more compoennts.

## Properties

<Props of={MML} />

## Basic usage

<Playground>
  <MML source="<mml>hi</mml>" />
</Playground>

## Basic usage with stateful components

<Playground>
  <MML
    source={`
      <mml name="windows_form" type="card">
        <image src="https://getstream.github.io/mml-react/public/window.jpg" />
        <number name="windows" value="12" />
        <button name="confirm" value="Confirm">Confirm</button>
      </mml>
    `}
  />
</Playground>

## Callback support

If onSubmit callback return a promise, it will be awaited.

<Playground>
  <MML
    onSubmit={(data) => console.info(data)}
    source={`
      <mml type="card">
        <input name="answer" label="Enter your phone number" placeholder="e.g. 999-999-9999"></input>
        <button name="action" value="submit">Submit</button>
      </mml>
    `}
  />
</Playground>

## Different use cases

<Playground>
  <MML
    source={`
      <mml>
        <text>Hi Jack, I'm painting John's house, looks like they also need some plumbing work, do any of these times work for you on Saturday?</text>
        <button_list>
          <button name="time" value="9:30">9:30</button>
          <button name="time" value="10:00">10:00</button>
          <button name="time" value="11:30">11:30</button>
          <button name="cancel" value="cancel">Cancel</button>
        </button_list>
      </mml>
    `}
  />
</Playground>

###

<Playground>
  <MML
    source={`
      <mml>
        <text>Hi Jack, I'm painting John's house, looks like they also need some plumbing work, do any of these times work for you on Saturday?</text>
        <button_list>
          <button name="time" value="9:30">9:30</button>
          <button name="time" value="10:00">10:00</button>
          <button name="time" value="11:30">11:30</button>
          <button name="cancel" value="cancel">Cancel</button>
        </button_list>
      </mml>
    `}
  />
</Playground>

## Invalid MML string

In case of an error, you can override default Error component

<Playground>
  <MML source="<mml><invalid /></mml>" Error={({ error }) => <div style={{ color: 'red' }}>{error}</div>} />
</Playground>

## With `card` type

Shows the mml content inside a card like UI, see the [`Number` component usage](/components/number) as a further example. It automatically wraps the mml components in a `<div class="mml-card">...</div>`.

<Playground>
  <MML
    source={`
      <mml type="card">
        <input name="answer" label="Enter your phone number" placeholder="e.g. 999-999-9999"></input>
        <button name="action" value="submit">Submit</button>
      </mml>
    `}
  />
</Playground>
