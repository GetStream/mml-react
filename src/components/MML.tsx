import React, { useState, useMemo, useCallback, FC, ComponentType, FormEvent } from 'react';

import { Parse } from '../parser';
import { ConvertorType } from '../converters';
import { MMLContext } from '../context';
import { Loader as LoaderComponent, LoaderProps } from './Loader';
import { Error as ErrorComponent, ErrorProps } from './Error';
import { Success as SuccessComponent, SuccessProps } from './Success';

export type MMLProps = {
  /** The MML string source to render */
  source: string;
  /** The convert config allows you to overwrite the MML to react conversion */
  converters?: Record<string, ConvertorType>;
  /** The submit callback whenever a form is submitted, submit is expected to return a promise */
  onSubmit?: (data: Record<string, unknown>) => Promise<any>;
  /** Custom classname, appended to wrapper classname */
  className?: string;
  /** The Loader component */
  Loader?: ComponentType<LoaderProps>;
  /** The error component */
  Error?: ComponentType<ErrorProps>;
  /** The success message component */
  Success?: ComponentType<SuccessProps>;
};

/**
 * MML root component
 */
export const MML: FC<MMLProps> = ({
  source,
  onSubmit,
  converters,
  className = '',
  Loader = LoaderComponent,
  Error = ErrorComponent,
  Success = SuccessComponent,
}) => {
  const [state, setState] = useState({});
  const [error, setError] = useState('');
  const [submitState, setSubmitState] = useState({ loading: false, error: '', success: '' });

  // expose helpers for form elements to change the state
  const setValue = useCallback((name: string, value: any) => {
    setState((prevState) => ({ ...prevState, [name]: value }));
  }, []);

  const handleSubmit = useCallback(
    async (event: FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      if (!onSubmit) return console.warn('missing submit handler');

      try {
        setSubmitState({ loading: true, error: '', success: '' });
        await onSubmit(state);
        setSubmitState({ loading: false, error: '', success: 'submitted' });
      } catch (e) {
        setSubmitState({ loading: false, error: 'something is broken', success: '' });
      }
    },
    [state, onSubmit],
  );

  const children = useMemo(() => {
    try {
      const tree = Parse(source, converters);
      setState(tree.initialState()); // get initial state for all input elements in MML
      return tree.toReact();
    } catch (e) {
      console.warn('mml parsing error: ', source, e);
      setError("This chat message has invalid formatting and can't be shown");
      return null;
    }
  }, [source, converters]);

  return (
    <div className={`mml-container mml-wrap ${className}`}>
      {error ? (
        <Error error={error} />
      ) : (
        <MMLContext.Provider value={{ ...state, submitState, setValue }}>
          <form onSubmit={handleSubmit}>
            {children}
            {submitState.loading && <Loader loading={submitState.loading} />}
            {submitState.success && <Success success={submitState.success} />}
            {submitState.error && <Error error={submitState.error} />}
          </form>
        </MMLContext.Provider>
      )}
    </div>
  );
};
