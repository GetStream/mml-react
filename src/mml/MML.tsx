import React, { useState, useMemo, useCallback, FC, ComponentType, FormEvent } from 'react';

import { Parse, ConvertorType } from '../parser';
import { Loading as LoadingComponent, LoadingProps } from '../components/Loading';
import { Error as ErrorComponent, ErrorProps } from '../components/Error';
import { Success as SuccessComponent, SuccessProps } from '../components/Success';

export type MMLProps = {
  /** The MML string source to render */
  source: string;
  /** The convert config allows you to overwrite the MML to react conversion */
  converters?: Record<string, ConvertorType>;
  /** The submit callback whenever a form is submitted, submit is expected to return a promise */
  onSubmit?: (data: Record<string, any>) => Promise<any> | any;
  /** Select a theme from built in available themes, it will be added as a className to the container(root) div */
  theme?:
    | 'base-dark'
    | 'messaging-light'
    | 'messaging-dark'
    | 'team-light'
    | 'team-dark'
    | 'commerce-light'
    | 'commerce-dark'
    | 'livestream-light'
    | 'livestream-dark'
    | string;
  /** Custom classname, appended to wrapper classname */
  className?: string;
  /** The Loading component, accepts null to render nothing */
  Loading?: ComponentType<LoadingProps> | null;
  /** The error component, accepts null to render nothing */
  Error?: ComponentType<ErrorProps> | null;
  /** The success message component, accepts null to render nothing */
  Success?: ComponentType<SuccessProps> | null;
};

/**
 * MML root component
 */
export const MML: FC<MMLProps> = ({
  source,
  onSubmit,
  converters,
  theme = '',
  className = '',
  Loading = LoadingComponent,
  Error = ErrorComponent,
  Success = SuccessComponent,
}) => {
  const [error, setError] = useState('');
  const [submitState, setSubmitState] = useState({ loading: false, error: '', success: '' });
  const tree = useMemo(() => {
    try {
      return Parse(source, converters);
    } catch (e) {
      console.warn('mml parsing error: ', source, e);
      setError("This chat message has invalid formatting and can't be shown");
      return null;
    }
  }, [source, converters]);

  const handleSubmit = useCallback(
    async (event: FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      const state: Record<string, any> = {};
      if (tree?.name) state.mml_name = tree.name; // include mml_name in the data
      new FormData(event.currentTarget).forEach((value, key) => {
        state[key] = value;
      });

      if (!onSubmit) return console.warn('Forgot to pass onSubmit prop to <MML/>? payload:', state);

      try {
        setSubmitState({ loading: true, error: '', success: '' });
        await onSubmit(state);
        setSubmitState({ loading: false, error: '', success: 'submitted' });
      } catch (e) {
        setSubmitState({ loading: false, error: 'Failed to send the request. Please try again.', success: '' });
      }
    },
    [onSubmit, tree],
  );

  return (
    <div className={`mml-container ${theme} ${className}`} data-testid="mml-container">
      {error ? (
        <div className="mml-wrap">{Error && <Error error={error} />}</div>
      ) : (
        <form onSubmit={handleSubmit} className="mml-wrap" data-testid="mml-form">
          {tree?.type ? <div className="mml-card">{tree?.reactElements}</div> : tree?.reactElements}
          {submitState.loading && Loading && <Loading loading={submitState.loading} />}
          {submitState.success && Success && <Success success={submitState.success} />}
          {submitState.error && Error && <Error error={submitState.error} />}
        </form>
      )}
    </div>
  );
};
