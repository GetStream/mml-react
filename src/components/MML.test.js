import React from "react";
import renderer from "react-test-renderer";
import { ParseMMLSource, XMLtoMMLTree } from "../parser";
import { MML } from "./MML";
import { examples } from "../examples";

test("mml name and simple text field", () => {
  const mml = '<mml name="john">hi</mml>';
  const nodes = ParseMMLSource(mml);
  const tree = XMLtoMMLTree(nodes);
  expect(tree.name).toEqual("john");
  expect(tree.children.length).toBe(1);
  const rTree = renderer.create(<MML source={mml} />).toJSON();
  expect(rTree).toMatchSnapshot();
});

test("simple carousel", () => {
  const mml = "<carousel><item>a</item><item>b</item></carousel>";
  const rTree = renderer.create(<MML source={mml} />).toJSON();
  expect(rTree).toMatchSnapshot();
});

test("simple input", () => {
  const mml = '<input name="name" value="John" />';
  const nodes = ParseMMLSource(mml);
  const tree = XMLtoMMLTree(nodes);
  const state = tree.initialState();
  console.log("state", state);
  expect(state).toEqual({ name: "John" });
  const rTree = renderer.create(<MML source={mml} />).toJSON();
  expect(rTree).toMatchSnapshot();
});

test("invalid input tag", () => {
  // note how name is missing
  const mml = '<input value="John" />';
  const nodes = ParseMMLSource(mml);
  const tree = XMLtoMMLTree(nodes);
  const errors = tree.validateTree();
  expect(errors.length).toBe(1);
  expect(errors[0]).toEqual("Attribute name is required for tag input");
});

test("input tags should have data", () => {
  const mml = '<input name="myinput" value="1" />';
  const nodes = ParseMMLSource(mml);
  const tree = XMLtoMMLTree(nodes);
  expect(tree.hasData()).toBe(true);
});

test("text tags should not have data", () => {
  const mml = "<text>hi</text>";
  const nodes = ParseMMLSource(mml);
  const tree = XMLtoMMLTree(nodes);
  expect(tree.hasData()).toBe(false);
});

test("invalid MML 1", () => {
  const mml = '<input name="test" value=1 />';
  const rTree = renderer.create(<MML source={mml} />).toJSON();
  expect(rTree).toMatchSnapshot();
});

test("invalid MML 2", () => {
  const mml =
    '<image src="https://turn5.scene7.com/is/image/Turn5/J20850?wid=810&hei=608&op_usm=0.8,1,10,0&amp;test=test" />';
  const rTree = renderer.create(<MML source={mml} />).toJSON();
  expect(rTree).toMatchSnapshot();
});

test.only("examples should work", () => {
  for (let example of examples) {
    let rTree = renderer.create(<MML source={example.mml} />).toJSON();
    expect(rTree).toMatchSnapshot();
  }
});

test("only supported attributes are allowed", () => {
  const mml = '<mml name="john">hi</mml>';
  const nodes = ParseMMLSource(mml);
  const tree = XMLtoMMLTree(nodes);
  expect(tree.name).toEqual("john");
  expect(tree.children.length).toBe(1);
  const rTree = renderer.create(<MML source={mml} />).toJSON();
  expect(rTree).toMatchSnapshot();
});
