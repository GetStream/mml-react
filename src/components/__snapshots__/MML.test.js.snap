// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid MML 2 1`] = `
<div
  className="mml-container"
/>
`;

exports[`mml name and simple text field 1`] = `
<div
  className="mml-container"
>
  <p>
    hi
  </p>
</div>
`;

exports[`only supported attributes are allowed 1`] = `
<div
  className="mml-container"
>
  <p>
    hi
  </p>
</div>
`;

exports[`simple carousel 1`] = `
<div
  className="mml-container"
>
  <div
    className="mml-carousel"
  >
    <div
      className="slick-slider slick-initialized"
    >
      
      <div
        className="slick-list"
      >
        <div
          className="slick-track"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "WebkitTransform": "translate3d(0px, 0px, 0px)",
              "WebkitTransition": "",
              "msTransform": "translateX(0px)",
              "opacity": 1,
              "transform": "translate3d(0px, 0px, 0px)",
              "transition": "",
            }
          }
        >
          <div
            aria-hidden={false}
            className="slick-slide slick-active slick-current"
            data-index={0}
            onClick={[Function]}
            style={
              Object {
                "outline": "none",
                "width": 0,
              }
            }
            tabIndex="-1"
          >
            <div>
              <div
                className="mml-carousel-item"
                style={
                  Object {
                    "display": "inline-block",
                    "width": "100%",
                  }
                }
                tabIndex={-1}
              >
                <p>
                  a
                </p>
              </div>
            </div>
          </div>
          <div
            aria-hidden={false}
            className="slick-slide slick-active"
            data-index={1}
            onClick={[Function]}
            style={
              Object {
                "outline": "none",
                "width": 0,
              }
            }
            tabIndex="-1"
          >
            <div>
              <div
                className="mml-carousel-item"
                style={
                  Object {
                    "display": "inline-block",
                    "width": "100%",
                  }
                }
                tabIndex={-1}
              >
                <p>
                  b
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      
      
    </div>
  </div>
</div>
`;

exports[`simple input 1`] = `
<div
  className="mml-container"
>
  <form
    onSubmit={[Function]}
  >
    <input
      name="name"
      onChange={[Function]}
      value="John"
    />
  </form>
</div>
`;
